lib/Mail/Milter/Authentication/Metric.pm

```diff
@@ -125,10 +125,21 @@ sub prom {
     if ( ! -e $metric_tempfile ) {
         $prom = undef;
     }
+    if ( ! -d $metric_tempfile ) {
+        # If metric_tempfile is a regular file then we need to re-init with a directory
+        # this is likely a restart after upgrade.
+        $prom = undef;
+    }
 
     if ( ! $prom ) {
+        if ( -f $metric_tempfile ) {
+            unlink $metric_tempfile;
+        }
+        if ( ! -d $metric_tempfile ) {
+            mkdir $metric_tempfile, 0700;
+        }
         $self->dbgout( 'Metrics', "Setup new metrics file $metric_tempfile", LOG_DEBUG );
-        $prom = eval{ Prometheus::Tiny::Shared->new(filename => $metric_tempfile, init_file => 1) };
+        $prom = eval{ Prometheus::Tiny::Shared->new(filename => $metric_tempfile.'/authmilter_metrics', init_file => 1) };
```

Comment:
Seems like you should probably log something if this fails, even if you intend to continue. But this is status quo anyway.

Reply:
We do,
handle_exception takes care of known errors (read timeouts) if the metrics object has a handler object, in those cases failure to create this object (being an unknown exception) will throw the exception up to the caller where it will be logged.
If we don't have a handler object (because we are working in the context of the main server, not a processing instance) then handle_exception does nothing. This happens when we are setting up the handler structure and registering the metrics, and also during some server management stages.

The following if $prom check falls through to the else, which logs the error.

Reply:
omg sorry, I completely missed the next line. Very good!